### ISSUE TITLE
futhark does not compile on windows

### ISSUE THREAD

**Issue by flip111** [diku-dk/futhark#832]
See 
[futhark_build.txt](https://github.com/diku-dk/futhark/files/4073976/futhark_build.txt) notable line:

```
C:\sr\snapshots\9bcf929c\lib\x86_64-windows-ghc-8.8.1\haskell-src-exts-1.22.0-620ASTx4cKYEetyTEBU7vQ\HShaskell-src-exts-1.22.0-620ASTx4cKYEetyTEBU7vQ.o: unhandled PEi386 relocation type 0
```

This looks to be similar to https://github.com/digital-asset/ghc-lib/issues/142

**Comment by athas** [comment]
Looks like this is fixed in GHC 8.8.2, which came out today.  I'll update the `stack.yaml` snapshot when one becomes available.

**Comment by flip111** [comment]
230b799f4905dee59511e4c0a14e593ef03f9f6a (16 Oct 2019) is the latest working version at the moment

**Comment by athas** [comment]
The new stack snapshot uses GHC 8.8.2.  Could you check whether it fixes it for you?

**Comment by flip111** [comment]
Unfortunately it's not fixed with the new nightly.

**Comment by athas** [comment]
Drat.  This is a GHC bug, and I really am not very familiar with its code (least of all its Windows runtime).  I suggest reporting it on the [GHC bug tracker](https://gitlab.haskell.org/ghc/ghc/issues).

**Comment by flip111** [comment]
There is a ticket for it here https://gitlab.haskell.org/ghc/ghc/issues/17599

**Comment by flip111** [comment]
Looks like a fix didn't get in 8.8.2 after all. Need to wait until the next release

**Comment by athas** [comment]
I see.  Thanks for looking into this!

**Comment by reversefold** [comment]
I also ran into this. The fix is supposedly in master of GHC 8.8 so should be in GHC 8.8.3.

In the meantime, no futhark version higher than 0.12.1 will compile on windows. (I just tested 0.12.1 and it does finish compiling at least.)

@athas Is there a way we could use a GHC 8.8.x nightly build or something to get around the issue? I'm not familiar with haskell, ghc, or stack myself, I'm new to this and trying out futhark.

**Comment by athas** [comment]
Futhark can as of recently also be built with `cabal`, although I have yet to document it.  If you manually install a sufficiently new GHC (that is, 8.8.3) and the `cabal` command line tool, you should be able to do `cabal install`.  By default, I think `cabal` will put the binaries somewhere nonobvious, but it will tell you where so you can fish them out.

I think installing the [core Haskell Platform](https://www.haskell.org/platform/windows.html) will get you everything you need.

**Comment by reversefold** [comment]
The Haskell Platform is still 8.6.5, so that doesn't really help unfortunately.

**Comment by athas** [comment]
Can you install the 8.8.3 [Windows GHC binary](https://www.haskell.org/ghc/download_ghc_8_8_3.html#windows64) directly?  It's also installable via [Chocolatey](https://chocolatey.org/).

In fact, I'm setting up some Windows CI on Azure pipelines right this moment, because I guess it's time we start taking it seriously again.  I use Chocolatey there.

**Comment by reversefold** [comment]
I tried getting ghc 8.8.82 and cabal 3.0 manually but then realized they were in chocolatey. Those installed fine.

Combining those with the haskell platform's mingw and msys (via path) I was able to get the build started but I'm still getting some errors. Oddly it's saying that futhark is 0.12.1, but this appears to be in futhark.cabal file.
```
PS C:\Users\paper\Dropbox\futhark-windows\futhark> cabal build
Warning: Requested index-state2020-02-05T00:00:00Z is newer than
'hackage.haskell.org'! Falling back to older state (2020-02-04T23:36:26Z).
Resolving dependencies...
Build profile: -w ghc-8.8.3 -O1
In order, the following will be built (use -v for more details):
 - futhark-0.12.1 (lib) (first run)
 - futhark-0.12.1 (exe:futhark) (first run)
Configuring library for futhark-0.12.1..
Preprocessing library for futhark-0.12.1..
cabal.exe: can't find source for Futhark\CodeGen\ImpGen\Kernels\SegGenRed in
src,
C:\Users\paper\Dropbox\futhark-windows\futhark\dist-newstyle\build\x86_64-windows\ghc-8.8.3\futhark-0.12.1\build\autogen,
C:\Users\paper\Dropbox\futhark-windows\futhark\dist-newstyle\build\x86_64-windows\ghc-8.8.3\futhark-0.12.1\build\global-autogen

cabal.exe: Failed to build futhark-0.12.1 (which is required by exe:futhark
from futhark-0.12.1).
```

**Comment by athas** [comment]
Bizarrely, that looks like it's using an old `.cabal` file (those are automatically generated from `package.yaml`).  I've been considering for a file whether to go back to explicit `.cabal` files, and I guess this pushesd me over the edge (especially since there is no `hpack` in Chocolatey).

Try the [use-cabal-file](https://github.com/diku-dk/futhark/tree/use-cabal-file) branch, which has an explicit `futhark.cabal`.

**Comment by reversefold** [comment]
This happened because I has previously checked out the v0.12.1 tag and tried to build. `git clean` wouldn't delete the file and I didn't see it in the `.gitignore` but I'm guessing those first few lines might have been ignoring it....

I deleted the `futhark.cabal`, modified `stack.yaml` to have `resolver: ghc-8.8.3` and tried some more combinations of `cabal build` and `stack build`, modifying `stack.yaml` to add a bunch of entries to `extra-deps` per the output.

Eventually I somehow made it into a state where `cabal build` worked for me and built a futhark.exe in `dist-newstyle\build\x86_64-windows\ghc-8.8.3\futhark-0.15.0\x\futhark\build\futhark\futhark.exe`

At this point I'm going to try to use this build (later today, though) and see if it works for me.

**Comment by athas** [comment]
Good you got it working.  Stack also just (today) released the first snapshot with GHC 8.8.3, which I have upgraded to in `master`.

**Comment by athas** [comment]
The Azure CI now also compiles Futhark on Windows (with `cabal` and GHC 8.10.1).  GHC also seems to be taking testing on Windows a bit more seriously, so hopefully the constant breakage will stop.
